src/cockpit/jsonutil.py

```diff
@@ -135,3 +135,23 @@ def create_object(message: 'JsonObject | None', kwargs: JsonObject) -> JsonObjec
         result[json_key] = value
 
     return result
+
+
+def json_merge_patch(current: JsonObject, patch: JsonObject) -> JsonObject:
+    """Perform a JSON merge patch (RFC 7396) using 'current' and 'patch'.
+    Neither of the original dictionaries is modified — the result is returned.
+    """
+    # Always take a copy ('result') — we never modify the input ('current')
+    result = dict(current)
+    for key, patch_value in patch.items():
+        if isinstance(patch_value, dict):
+            current_value = current.get(key, None)
+            if not isinstance(current_value, dict):
+                current_value = {}
```

Comment:
I don't understand this. Either the entries of `current` are all `dict`s - then this should be reflected in a more specific type signature, and you can simplify to `.get(key, {})`; but that seems implausible to me, e.g. package manifests certainly have string or int keys at the top level?

Or the keys can have different types, in which case you'd entirely ignore them in the recursion call below.

Reply:
In other words, this assumes that the types in `patch` always match the types in `current`. This is probably ok for the current code, but it still feels like a trap. Perhaps replace this with an assertion that `current_value` is a `dict` or `None`?

Reply:
The idea here is that, in a merge patch, the rules go something like so:

if patch is non-dict value:
  old value is deleted, new value is patch value

else (patch is dict value):
  if old value is dict:
    patch recursively
  else if old value is not dict:
    patch recursively, based on an empty start ({})

That last line is what we're doing here.  eg: it may indeed be the case that you start with:

{
  "x": "y"
}

and get

{
  "x": {"a": "b"}
}

(ie: the type of "x" changes).

BUT: although the code works in practice (since we never have any mapping type that's not a dictionary) the code is not properly implementing the interface: we shouldn't be checking for `dict` here but rather any `Mapping`.

```python
>>> isinstance({}, typing.Mapping)
True
```

so I'll fix this.

Reply:
I see now, sorry for being dense. But I'm glad that it triggered that little improvement. :grin: